File: Problems & Solutions/S/Select Nodes/30 January 2023/Select Nodes.py
class Solution:
    
    def dfs(self, adj, src, par, val):
        for v in adj[src]:
            if v!=par:
                self.dfs(adj,v,src,val)
        
        for v in adj[src]:
            if v!=par:
                val[src][0] += val[v][1];
                val[src][1] += min(val[v][1],val[v][0])

    def countVertex(self, N, edges):
        adj=[[] for i in range(N+1)]
        for i in range(N-1):
            adj[edges[i][0]].append(edges[i][1])
            adj[edges[i][1]].append(edges[i][0]) 
        
        val=[[] for i in range(N+1)]
        for i in range(1,N+1):
            val[i] = [0,1]
        
        self.dfs(adj,1,1,val)
        
        return min(val[1][0],val[1][1])

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__': 
    t = int (input ())
    for _ in range (t):
        N = int(input())
        edges=[]
        for _ in range(N-1):
            arr = list(map(int,input().split()))
            edges.append(arr)

        ob = Solution()
        print(ob.countVertex(N, edges))
# } Driver Code Ends

File: Problems & Solutions/S/Select Nodes/30 January 2023/Select Nodes.cpp
//{ Driver Code Starts
//Initial Template for C++

#include<bits/stdc++.h> 
using namespace std; 

// } Driver Code Ends
//User function Template for C++

class Solution{
  public:
  
    vector<int> vis;
    int count = 0;
    bool dfs(vector<int> adj[], int i){
        vis[i] = 1;
        bool select = false;
        
        for (auto j:adj[i]){
            
            if(!vis[j] && !dfs(adj, j)) select = true;
        }
        if (select){
            count++;
        }
        return select;
    }
    
    int countVertex(int N, vector<vector<int>>edges){
        
        vector<int> adj[N+1];
        vis = vector<int>(N+1, 0);
        
        for(auto i:edges){
            adj[i[0]].push_back(i[1]);
            adj[i[1]].push_back(i[0]);
        }
        
        dfs(adj, 1);
        return count;
    }
};

//{ Driver Code Starts.

int main() 
{ 
    int t;
    cin>>t;
    while(t--)
    {
        int N;
        cin >> N;
        vector<vector<int>>edges;
        for(int i = 0; i < N-1; i++){
            int x,y;
            cin >> x >> y;
            vector<int> edge;
            edge.push_back(x);
            edge.push_back(y);
            edges.push_back(edge);
        }
        
        Solution ob;
        cout << ob.countVertex(N, edges) << endl;
    }
    return 0; 
} 
// } Driver Code Ends
